// graphql-language-service-server/esm/constants.js
export const PARSER_OPTIONS = {
    allowImportExportEverywhere: true,
    allowReturnOutsideFunction: true,
    allowSuperOutsideMethod: true,
    allowAwaitOutsideFunction: true,
    errorRecovery: true,
    sourceType: 'module',
    strictMode: false,
};
export const DEFAULT_SUPPORTED_GRAPHQL_EXTENSIONS = [
    '.graphql',
    '.graphqls',
    '.gql',
];
export const TAG_MAP = {
    graphql: true,
    gql: true,
    graphqls: true,
};
export const DEFAULT_SUPPORTED_EXTENSIONS = [
    '.js',
    '.cjs',
    '.mjs',
    '.es',
    '.esm',
    '.es6',
    '.ts',
    '.jsx',
    '.tsx',
    '.vue',
    '.svelte',
    '.astro',
    '.cts',
    '.mts',
];
export const BABEL_PLUGINS = [
    'asyncDoExpressions',
    'asyncGenerators',
    'bigInt',
    'classProperties',
    'classPrivateProperties',
    'classPrivateMethods',
    'classStaticBlock',
    'doExpressions',
    'decimal',
    'decorators-legacy',
    'destructuringPrivate',
    'dynamicImport',
    'exportDefaultFrom',
    'exportNamespaceFrom',
    'functionBind',
    'functionSent',
    'importMeta',
    'importAssertions',
    'jsx',
    'logicalAssignment',
    'moduleBlocks',
    'moduleStringNames',
    'nullishCoalescingOperator',
    'numericSeparator',
    'objectRestSpread',
    'optionalCatchBinding',
    'optionalChaining',
    'privateIn',
    'regexpUnicodeSets',
    'throwExpressions',
    'topLevelAwait',
];
